File: apps/ios/modules/volume/ios/ValenceVolumeModule.swift
import AVFoundation
import ExpoModulesCore
import MediaPlayer

/// The phone's volume, for a slider drawn like the rest of the player.
///
/// iOS only lets an app set the volume through the system's own slider, so a
/// hidden `MPVolumeView` is kept on screen and its slider is moved on our
/// behalf. Keeping one on screen also stops iOS drawing its volume box over the
/// player when the buttons on the side of the phone are pressed. The volume is
/// read from the audio session, and watched there so the slider follows those
/// buttons.
public class ValenceVolumeModule: Module {
  private var watching: NSKeyValueObservation?

  public func definition() -> ModuleDefinition {
    Name("ValenceVolume")

    Events("onVolume")

    Function("now") { () -> Float in
      AVAudioSession.sharedInstance().outputVolume
    }

    AsyncFunction("set") { (to: Float) in
      ValenceVolumeView.setVolume(to)
    }.runOnQueue(.main)

    OnStartObserving {
      self.watching = AVAudioSession.sharedInstance().observe(\.outputVolume, options: [.new]) {
        [weak self] session, _ in
        self?.sendEvent("onVolume", ["volume": session.outputVolume])
      }
    }

    OnStopObserving {
      self.watching?.invalidate()
      self.watching = nil
    }

    View(ValenceVolumeView.self) {}
  }
}

/// Holds the system's volume slider, all but invisible.
public class ValenceVolumeView: ExpoView {
  private static weak var shown: ValenceVolumeView?

  private let volume = MPVolumeView()

  public required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)

    isUserInteractionEnabled = false
    volume.alpha = 0.01
    addSubview(volume)
    ValenceVolumeView.shown = self
  }

  public override func didMoveToWindow() {
    super.didMoveToWindow()

    if window != nil {
      ValenceVolumeView.shown = self
    }
  }

  public override func layoutSubviews() {
    super.layoutSubviews()

    volume.frame = bounds
  }

  /// Moves the system's slider, which sets the volume of whatever is playing.
  static func setVolume(_ to: Float) {
    guard let slider = shown?.volume.subviews.compactMap({ $0 as? UISlider }).first else {
      return
    }

    slider.setValue(min(max(to, 0), 1), animated: false)
    slider.sendActions(for: .valueChanged)
  }
}

File: apps/mobile/modules/volume/android/src/main/java/app/valence/modules/volume/ValenceVolumeModule.kt
package app.valence.modules.volume

import android.content.Context
import android.database.ContentObserver
import android.media.AudioManager
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlin.math.roundToInt

/**
 * The phone's media volume, for a slider drawn like the rest of the player.
 *
 * Android lets an app set it directly, so there is no hidden system slider as on an iPhone. It is
 * watched through the system settings, which change whenever the buttons on the side are pressed,
 * so the slider follows them.
 */
class ValenceVolumeModule : Module() {
  private val main = Handler(Looper.getMainLooper())
  private var watching: ContentObserver? = null

  private val audio: AudioManager?
    get() = appContext.reactContext?.getSystemService(Context.AUDIO_SERVICE) as AudioManager?

  override fun definition() = ModuleDefinition {
    Name("ValenceVolume")

    Events("onVolume")

    Function("now") {
      loudness()
    }

    AsyncFunction("set") { to: Double ->
      val manager = audio ?: return@AsyncFunction
      val most = manager.getStreamMaxVolume(AudioManager.STREAM_MUSIC)

      manager.setStreamVolume(
        AudioManager.STREAM_MUSIC,
        (to.coerceIn(0.0, 1.0) * most).roundToInt(),
        0,
      )
    }

    OnStartObserving {
      val observer = object : ContentObserver(main) {
        override fun onChange(selfChange: Boolean) {
          sendEvent("onVolume", mapOf("volume" to loudness()))
        }
      }

      appContext.reactContext?.contentResolver?.registerContentObserver(
        Settings.System.CONTENT_URI,
        true,
        observer,
      )
      watching = observer
    }

    OnStopObserving {
      watching?.let { appContext.reactContext?.contentResolver?.unregisterContentObserver(it) }
      watching = null
    }
  }

  /** How loud the media volume is, from nothing to everything. */
  private fun loudness(): Double {
    val manager = audio ?: return 1.0
    val most = manager.getStreamMaxVolume(AudioManager.STREAM_MUSIC)

    return if (most == 0) 1.0 else manager.getStreamVolume(AudioManager.STREAM_MUSIC).toDouble() / most
  }
}
